optimize/global.go

```diff
@@ -184,158 +177,169 @@ func minimizeGlobal(p *Problem, method GlobalMethod, settings *Settings, stats *
+	//   conditions in updating the stats and checking convergence. This also
+	//   ensures that MajorIteration updates happen in a predictable order.
+	// - When a termination condition is met, a Status is sent on statusChan.
+	//   only the first such termination signal is sent. This triggers a
+	//   PostIteration to be sent to Method, and triggers a series of shutdown steps.
 
-		// Evaluate location and/or update stats.
-		status := g.globalOperation(op, loc, x)
-		if status != NotTerminated {
-			break
-		}
-	}
-}
+	var (
+		finalStatus Status
+		finalError  error
+	)
 
-// globalStatus coordinates access to information shared between concurrently
-// executing optimization tasks.
-type globalStatus struct {
-	mux       *sync.RWMutex
-	stats     *Stats
-	status    Status
-	p         *Problem
-	startTime time.Time
-	optLoc    *Location
-	settings  *Settings
-	method    GlobalMethod
-	statuser  Statuser
-	err       error
-}
+	operations := make(chan GlobalTask, nTasks) // GlobalMethod sends tasks.
+	results := make(chan GlobalTask, nTasks)    // return results to GlobalMethod.
+	workerChan := make(chan GlobalTask)         // Delegate tasks to the workers.
+	evalStatsChan := make(chan GlobalTask)      // Send evaluation updates.
+	statusSent := make(chan struct{})           // Communicate the optimization is done.
 
-// getStatus returns the current status of the optimization.
-func (g *globalStatus) getStatus() Status {
-	var status Status
-	g.mux.RLock()
-	defer g.mux.RUnlock()
-	status = g.status
-	return status
-}
-
-func (g *globalStatus) incrementMajorIteration() {
-	g.mux.Lock()
-	defer g.mux.Unlock()
-	g.stats.MajorIterations++
-}
-
-func (g *globalStatus) updateOptLoc(loc *Location) {
-	g.mux.Lock()
-	defer g.mux.Unlock()
-	copyLocation(g.optLoc, loc)
-}
-
-// checkConvergence checks the convergence of the global optimization and returns
-// the status
-func (g *globalStatus) checkConvergence() Status {
-	g.mux.RLock()
-	defer g.mux.RUnlock()
-	return checkConvergence(g.optLoc, g.settings, false)
-}
-
-// updateStats updates the evaluation statistics for the given operation.
-func (g *globalStatus) updateStats(op Operation) {
-	g.mux.Lock()
-	defer g.mux.Unlock()
-	updateEvaluationStats(g.stats, op)
-}
+	// Launch the method.
+	go func() {
+		tasks := make([]GlobalTask, nTasks)
+		for i := range tasks {
+			tasks[i].Location = newLocation(dim, method)
+		}
+		method.RunGlobal(operations, results, tasks)
+	}()
 
-// updateStatus updates the status and error fields of g. This update only happens
-// if status == NotTerminated, so that the first different status is the one
-// maintained.
-func (g *globalStatus) updateStatus(s Status, err error) {
-	g.mux.Lock()
-	defer g.mux.Unlock()
-	if s != NotTerminated {
-		g.status = s
-		g.err = err
+	// Launch the workers. After workerChan is closed, each worker signals
+	// to the stats updater that they have successfully terminated.
+	for worker := 0; worker < nTasks; worker++ {
+		go func() {
```

Comment:
I'd assign the function literal to a variable and have the comment `After workerChan is closed ...` before it. Commenting `// Launch the workers.` would then be unnecessary.